app/expenses/crud.py

```diff
@@ -94,3 +100,50 @@ def generate_visualization(db, month):
     image_stream.seek(0)
 
     return image_stream
+
+
+def generate_report(
+    db: Session,
+    category: str | None,
+    start_date: date | None,
+    end_date: date | None
+):
+    query = db.query(Expense)
+
+    # Filtr: kategoria
+    if category:
+        query = query.filter(Expense.category == category)
+
+    # Filtr: data początkowa
+    if start_date:
+        query = query.filter(Expense.created_at >= start_date)
+
+    # Filtr: data końcowa
+    if end_date:
+        query = query.filter(Expense.created_at <= end_date)
+
+    try:
+        expenses = query.all()
+    except SQLAlchemyError as e:
+        raise ValueError(f"Database error: {str(e)}")
+
+    if not expenses:
+        raise_not_found()
```

Comment:
You are mixing how `crud.py` module raising exception. 

Recommended way is to raise custom exception like ValueError, ItemNotFoundError and then on routers.py level raise HTTPException or similar exception.